### PR TITLE
Bluetooth: Host: Ensure only connected peers affect CCC value

### LINKED ISSUE #81118: Bluetooth: Host: `_bt_gatt_ccc.value` may include disconnected peers
When `CONFIG_BT_SETTINGS_CCC_LAZY_LOADING=n`, the value of `_bt_gatt_ccc.value` is affected by CCC values from all bonds, not just connected peers.

The documentation for `_bt_gatt_ccc.value` specifies that only connected peers contribute to that value.

https://github.com/zephyrproject-rtos/zephyr/blob/67980643d799f76aff93118346a8dabfb472a642/include/zephyr/bluetooth/gatt.h#L939-L940

This bug makes the behavior of `_bt_gatt_ccc.value` depend on the state of `CONFIG_BT_SETTINGS_CCC_LAZY_LOADING`, so it cannot reasonably be fixed by keeping behavior and updating documentation.

### PR DESCRIPTION
**Pull request by HaavardRei**
Only consider connected peers when summing CCC values instead of
connected or bonded.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/81118

### CODE REVIEW

**Comment by HaavardRei** [comment]
Picks up the thread of two of the commits from https://github.com/zephyrproject-rtos/zephyr/pull/44136

**Review by jhedberg — DISMISSED** [review]

**Review by alwa-nordic — DISMISSED** [review]
I would have liked if I was kept as author or co-author of the commits, but no matter. The important thing is it gets in :)

The two commits should ideally come in two PRs, since they are two separate behavior changes.

**Comment by HaavardRei** [comment]
> I would have liked if I was kept as author or co-author of the commits, but no matter. The important thing is it gets in :)
> 
> The two commits should ideally come in two PRs, since they are two separate behavior changes.

Added you as co-author :) 
I think the two separate commits are sufficient, but can fix if it's important!

File: subsys/bluetooth/host/gatt.c
Comment (Thalley) [inline subsys/bluetooth/host/gatt.c:2204]: I think this has been attempted before. 

Isn't it dangerous to use values that are not defined as a bitfield in the BT core spec as a bitfield?

For example if the BT core spec ever introduces a new value 0x0003 then `value` here would be the same as `BT_GATT_CCC_NOTIFY | BT_GATT_CCC_INDICATE`. 

This change feels less future-proof than the existing, and does not fix any issues. The value here would also be 0x0003 which is not defined in Zephyr, and if applications are testing for `== BT_GATT_CCC_NOTIFY` or `== BT_GATT_CCC_INDICATE` (which we have several cases of in Zephyr), then those checks will break. 

The real issue is that the `cfg_changed` callback does not consider that individual clients have different values. 

```c
	/** @brief CCC attribute changed callback
	 *
	 *  @param attr   The attribute that's changed value
	 *  @param value  New value
	 */
	void (*cfg_changed)(const struct bt_gatt_attr *attr, uint16_t value);
```

should really be
```c
	/** @brief CCC attribute changed callback
	 *
	 *  @param conn   The connection that updated it's CCC value
	 *  @param attr   The attribute that's changed value
	 *  @param value  New value
	 */
	void (*cfg_changed)(struct bt_conn *conn, const struct bt_gatt_attr *attr, uint16_t value);
```

I would suggest to leave the existing callback as is, and introduce the new callback (and possibly deprecate the old one). 

It feels like this is trying to fix a fundamental issue by applying a band-aid, while also introducing a breaking change as the callback value of 0x0003 (BT_GATT_CCC_NOTIFY | BT_GATT_CCC_INDICATE) is undefined.

**Review by Thalley — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/gatt.c
Comment (alwa-nordic) [reply subsys/bluetooth/host/gatt.c:2204]: [It is a bit field.](https://www.bluetooth.com/wp-content/uploads/Files/Specification/HTML/Core-60/out/en/host/generic-attribute-profile--gatt-.html#:~:text=Table%203.11:%20Client%20Characteristic%20Configuration-,bit%20field,-definition)

Comment (HaavardRei) [reply subsys/bluetooth/host/gatt.c:2204]: > I think this has been attempted before.
> 
> Isn't it dangerous to use values that are not defined as a bitfield in the BT core spec as a bitfield?
> 
> For example if the BT core spec ever introduces a new value 0x0003 then `value` here would be the same as `BT_GATT_CCC_NOTIFY | BT_GATT_CCC_INDICATE`.

But it _is_ defined as a bitfield (Core 6.0):
image

Comment (Thalley) [reply subsys/bluetooth/host/gatt.c:2204]: I stand corrected :) 

Thanks. 
Should we consider redefining the values in Zephyr from `0x0001` and 0x0002` to `BIT(0)` and `BIT(2)` then?

In Zephyr we have quite a few cases where we treat it not as a bitfield: 
image
image

They should be updated to use `&` instead of `==` then, as otherwise if the GATT clients subscribe to both indication and notification, then those checks will now fail (granted that all the `== BT_GATT_CCC_NOTIFY` would fail anyways due to the current version using the max of the 2. 

I would also prefer if we _somewhere_ (either in the values or in the documentation for `cfg_changed` or somewhere else) make it clear that these are now bitfields in Zephyr, which they have not been so far. 

Not sure if this change counts as a breaking stable API change, as they have been treated as non-bitfield for 9+ years in Zephyr. I'll leave that up to the maintainers of the host to decide.

Comment (HaavardRei) [reply subsys/bluetooth/host/gatt.c:2204]: > Thanks. Should we consider redefining the values in Zephyr from `0x0001` and 0x0002`to`BIT(0)`and`BIT(2)` then?

I agree! And I'll look into switching the checks to bitfields. 
As for the breaking/nonbreaking, we should at least have a release note/changelog. Thoughts @alwa-nordic?

Comment (alwa-nordic) [reply subsys/bluetooth/host/gatt.c:2204]: > This change feels less future-proof than the existing, and does not fix any issues. The value here would also be 0x0003 which is not defined in Zephyr, and if applications are testing for `== BT_GATT_CCC_NOTIFY` or `== BT_GATT_CCC_INDICATE` (which we have several cases of in Zephyr), then those checks will break.

Yes, this breaks API. But it breaks only in the case when one client enabled indications, and another has enabled notifications. (And the strange case when the same client has enabled both.) The Characteristic Declaration will specify which one of those is supported. Do we have examples of services that support both? Of course there is the issue of a DOS by a rouge client, but the code that does `== BT_GATT_CCC_NOTIFY` will already be vulnerable to this, since we don't check the characteristic properties before we accept the write to the CCC.

> The real issue is that the `cfg_changed` callback does not consider that individual clients have different values. \[We should add `attr` to `cfg_changed`].

Yes, I agree there a need for an event for when `_bt_gatt_ccc.cfg` changes. But this is an orthogonal issue, right? `_bt_gatt_ccc.cfg_changed` is the event for when `_bt_gatt_ccc.value` changes.

This PR instead includes a change to the definition of `_bt_gatt_ccc.value` so it makes sense when a characteristic support both indication and notification. Does this have any real-world uses? I don't know. But the spec allows it, AFAIK.

Comment (alwa-nordic) [reply subsys/bluetooth/host/gatt.c:2204]: Good examples. In that case, this really is a breaking change. To be on the safe side, we should go with a deprecation, would you agree?

Comment (alwa-nordic) [reply subsys/bluetooth/host/gatt.c:2204]: We should remove the problematic change from this PR and evaluate how to move forward. I think we should either leave it as it is, or deprecate it without replacement. And we should, as @Thalley points out, address the lack of an event for changes in subscription. The only way to do that without breaking API is to add a new callback.

Comment (alwa-nordic) [reply subsys/bluetooth/host/gatt.c:2204]: Adding issues:

 - https://github.com/zephyrproject-rtos/zephyr/issues/81004
 - https://github.com/zephyrproject-rtos/zephyr/issues/81005

Comment (HaavardRei) [reply subsys/bluetooth/host/gatt.c:2204]: > We should remove the problematic change from this PR and evaluate how to move forward. 

Meaning keeping only the commit that checks whether the peer is connected or not?

**Review by alwa-nordic — APPROVED** [review]

**Review by Thalley — APPROVED** [review]

File: subsys/bluetooth/host/gatt.c
Comment (alwa-nordic) [reply subsys/bluetooth/host/gatt.c]: I added another issue:
 - https://github.com/zephyrproject-rtos/zephyr/issues/81056

**Review by jhedberg — APPROVED** [review]

**Pull request merged**
